solarforecastarbiter/validation/validator.py

```diff
@@ -0,0 +1,189 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Feb 15 14:08:20 2019
+
+@author: cwhanse
+"""
+
+from pvlib.tools import cosd
+import numpy as np
+
+QCRAD_LIMITS = {'ghi_ub': {'mult': 1.5, 'exp': 1.2, 'min': 100},
+                'dhi_ub': {'mult': 0.95, 'exp': 1.2, 'min': 50},
+                'dni_ub': {'mult': 1.0, 'exp': 0.0, 'min': 0},
+                'ghi_lb': -4, 'dhi_lb': -4, 'dni_lb': -4}
+
+QCRAD_CONSISTENCY = {
+    'ghi_ratio':
+        {'high_zenith':
+            {'zenith_bounds': [0.0, 75], 'ghi_bounds': [50, np.Inf],
+             'ratio_bounds': [0.85, 1.15]},
+         'low_zenith':
+            {'zenith_bounds': [75, 93], 'ghi_bounds': [50, np.Inf],
+             'ratio_bounds': [0.92, 1.08]}},
+    'dhi_ratio':
+        {'high_zenith':
+            {'zenith_bounds': [0.0, 75], 'ghi_bounds': [50, np.Inf],
+             'ratio_bounds': [0.0, 1.10]},
+         'low_zenith':
+            {'zenith_bounds': [75, 93], 'ghi_bounds': [50, np.Inf],
+             'ratio_bounds': [0.0, 1.05]}}}
+
+
+def _apply_limit(val, lb=None, ub=None):
+    if (lb is not None) & (ub is not None):
+        return (val > lb) & (val < ub)
+    elif lb is not None:
+        return (val > lb)
+    elif ub is not None:
+        return (val < ub)
+    else:
+        pass
+
+
+def _QCRad_ub(dni_extra, sza, coeff):
+    return coeff['mult'] * dni_extra * cosd(sza)**coeff['exp'] + coeff['min']
+
+
+def check_irradiance_limits_QCRad(irrad, coeff=QCRAD_LIMITS):
+    """
+    Applies limits using the QCRad criteria.
+
+    Parameters:
+    -----------
+    irrad : DataFrame
+        ghi : float
+            Global horizontal irradiance in W/m^2
+        solar_zenith : float
+            Solar zenith angle in degrees
+        dni_extra : float
+            Extraterrestrial normal irradiance in W/m^2
+        dhi : float
+            Diffuse horizontal irradiance in W/m^2
```

Comment:
my reading of the `try/except` statements is that `dhi` and `dni` are supposed to be optional and the function will silently move on if they do not exist. If so, this should be documented. 

Related, I wonder if we should include a kwarg like `require_dni_dhi=False`. If we're trying to validate DNI data then we want to make sure it works.

Reply:
Good point, I'll add two Boolean arguments.